pkgs/native_assets_builder/lib/src/utils/file.dart

```diff
@@ -50,4 +51,99 @@ extension DirectoryExtension on Directory {
     }
     return last;
   }
+
+  /// Run [callback] with this Dart process having exclusive access to this
+  /// [Directory].
+  ///
+  /// Note multiple isolates and isolate groups in the same Dart process share
+  /// locks, so these will be able to enter the exclusive section simultanously.
+  ///
+  /// The required [timeout] parameter determines how long to retry before
+  /// giving up. If the [timeout] is exceeded a [TimeoutException] is thrown.
+  // TODO(dacoharkes): Can we also make it in-process safe?
+  // TODO(dacoharkes): Where do we place this code so hook writers can reuse it?
+  // I don't really want to add it to package:native_assets_cli. A new package
+  // in dart-lang/tools?
+  Future<T> exclusive<T>(
+    Future<T> Function() callback, {
+    required Duration timeout,
+  }) async {
+    const lockFileName = '.lock';
+    final lockFile = File.fromUri(uri.resolve(lockFileName));
+    final fileSystemExceptions = <FileSystemException>[];
+    await _ensureLockfileExists(lockFile, fileSystemExceptions);
+
+    final timeoutCompleter = Completer<void>();
+    var timedOut = false;
+    final timeoutTimer = Timer(timeout, () {
+      timedOut = true;
+      timeoutCompleter.complete();
+    });
+
+    final randomAccessFile = await lockFile.open(mode: FileMode.write);
+    await lockFile.writeAsString(
+      'Last acquired by ${Platform.resolvedExecutable} '
+      'running ${Platform.script} on ${DateTime.now()}.',
+    );
+    var ranCallback = false;
+    T? result;
+    final lockAndCallbackFuture =
+        randomAccessFile.lock(FileLock.blockingExclusive).then((lock) async {
+      if (!timedOut) {
+        // Lock was acquired before timeout.
+        // Cancel timer, otherwise the Dart process doesn't exit.
+        timeoutTimer.cancel();
+        try {
+          result = await callback();
+          ranCallback = true;
+        } finally {
+          await lock.unlock();
+        }
+      } else {
+        // Timed out, but lock was acquired.
+        await lock.unlock();
+        // TODO(dacoharkes): The current approach probably prevents the Dart
+        // process from exiting while waiting on the file lock after the timer
+        // has already timed out.
+        // The alternative would be to repeatedly try locking with
+        // `FileLock.exclusive` with an exponential back off. But the downside
+        // of that is more latency due to polling.
+        // Why don't we have a cancallable blocking locking API?
```

Comment:
That seems like a bad thing, no?

It's a leaky abstraction: It claims that it can acquire a lock within a certain time and result in `TimeoutException` if the time expired  - but actually it leaves the pending locking operation running as a side-effect. That side-effect may even prevent the application from shutting down ?!

A polling strategy doesn't have to use an exponential backoff (why?) - so it wouldn't have a huge latency increase either. If another process holds the lock, then chances are high it will take O(seconds) for it to release it. So if this code checks every second (or even every 100 ms), it's not really a problem.